File: training-data/perl/shunting_yard.pl
use strict;
use warnings;

my %PREC = ('+' => [1, 'L'], '-' => [1, 'L'], '*' => [2, 'L'], '/' => [2, 'L'], '^' => [3, 'R']);

sub to_rpn {
    my @tokens = shift =~ m{(\d+(?:\.\d+)?|[-+*/^()])}g;
    my (@out, @ops);
    for my $t (@tokens) {
        if ($t =~ /^\d/) { push @out, $t }
        elsif ($t eq '(') { push @ops, $t }
        elsif ($t eq ')') {
            push @out, pop @ops while @ops && $ops[-1] ne '(';
            die "mismatched parentheses\n" unless @ops;
            pop @ops;
        } else {
            my ($p, $assoc) = @{ $PREC{$t} };
            while (@ops && $ops[-1] ne '(') {
                my $top = $PREC{ $ops[-1] }[0];
                last unless $top > $p || ($top == $p && $assoc eq 'L');
                push @out, pop @ops;
            }
            push @ops, $t;
        }
    }
    die "mismatched parentheses\n" if grep { $_ eq '(' } @ops;
    return @out, reverse @ops;
}

sub eval_rpn {
    my @st;
    for my $t (@_) {
        if ($t =~ /^\d/) { push @st, $t; next }
        my ($b, $a) = (pop @st, pop @st);
        push @st, $t eq '+' ? $a + $b : $t eq '-' ? $a - $b : $t eq '*' ? $a * $b : $t eq '/' ? $a / $b : $a**$b;
    }
    return $st[0];
}

for my $e ('3 + 4 * 2 / (1 - 5) ^ 2 ^ 3', '(1 + 2) * (3 + 4)', '2 ^ 3 ^ 2') {
    my @rpn = to_rpn($e);
    print "$e => @rpn = ", eval_rpn(@rpn), "\n";
}

File: training-data/typescript/shunting_yard.ts
type Op = "+" | "-" | "*" | "/" | "^";
const PREC: Record<Op, number> = { "+": 1, "-": 1, "*": 2, "/": 2, "^": 3 };
const APPLY: Record<Op, (a: number, b: number) => number> = {
  "+": (a, b) => a + b,
  "-": (a, b) => a - b,
  "*": (a, b) => a * b,
  "/": (a, b) => a / b,
  "^": (a, b) => a ** b,
};
const isOp = (t: string): t is Op => t in PREC;

export function toRpn(tokens: string[]): string[] {
  const out: string[] = [];
  const ops: string[] = [];
  for (const t of tokens) {
    if (isOp(t)) {
      for (let top = ops.at(-1); top !== undefined && isOp(top) && (PREC[top] > PREC[t] || (PREC[top] === PREC[t] && t !== "^")); top = ops.at(-1)) {
        out.push(ops.pop()!);
      }
      ops.push(t);
    } else if (t === "(") ops.push(t);
    else if (t === ")") {
      while (ops.length && ops.at(-1) !== "(") out.push(ops.pop()!);
      if (!ops.pop()) throw new SyntaxError("mismatched parentheses");
    } else out.push(t);
  }
  while (ops.length) {
    const op = ops.pop()!;
    if (op === "(") throw new SyntaxError("mismatched parentheses");
    out.push(op);
  }
  return out;
}

export function evalRpn(rpn: string[]): number {
  const st: number[] = [];
  for (const t of rpn) {
    if (isOp(t)) {
      const b = st.pop()!, a = st.pop()!;
      st.push(APPLY[t](a, b));
    } else st.push(Number(t));
  }
  return st[0];
}

const rpn = toRpn("3 + 4 * 2 / ( 1 - 5 ) ^ 2 ^ 3".split(" "));
console.log(rpn.join(" "), "=", evalRpn(rpn));
